### PR TITLE
feat(pinpoint): Targeting half of Analytics migration

### PR DESCRIPTION
**Pull request by tjleing**
- [x] PR title and description conform to [Pull Request](https://github.com/aws-amplify/amplify-android/blob/main/CONTRIBUTING.md#pull-request-guidelines) guidelines.

*Issue #, if available:*

*Description of changes:*
Moved files over from Android SDK targeting folder, translated into Kotlin, attempted to clean up logic.

I'm not happy with the fact that `EndpointProfileLocation` and similar classes need getters/setters, but I believe it's necessary for backwards compatibility.  Open to having a discussion about that, since users will already need to migrate to some extent.

Missing: tests, splitting off push notifications cleanly (will need to be into core to communicate attributes like `channeltype`, `optout`, etc. into the endpoint)

*How did you test these changes?*
(Please add a line here how the changes were tested)
not tested yet, draft PR

- [ ] Added Unit Tests
- [ ] Tested on frontend (Add Screenshots)
- [ ] Successful logs (Attach them to the PR)

*Documentation update required?*
- [ ] No
- [ ] Yes (Please include a PR link for the documentation update)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/models/AndroidDeviceDetails.kt [hunk 1]
```diff
@@ -18,11 +18,11 @@ package com.amplifyframework.analytics.pinpoint.models
 import android.os.Build
 import java.util.Locale
 
-internal data class AndroidDeviceDetails(
+data class AndroidDeviceDetails(
     val carrier: String? = null,
-    val platformVersion: String? = Build.VERSION.RELEASE,
```
Comment (sdhuka) [inline aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/models/AndroidDeviceDetails.kt:23]: verify tests are passing with non-build system properites

File: aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/targeting/endpointProfile/EndpointProfile.kt [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+     * [EndpointProfile], the call may be ignored.
+     *
+     * @param name   The name of the custom attribute. The name will be truncated if it
+     * exceeds 50 characters.
+     * @param values An array of values of the custom attribute. The values will be truncated if
+     * it exceeds 100 characters.
+     * @return The same [EndpointProfile] instance is returned to allow for
+     * method chaining.
+     */
+    fun withAttribute(name: String?, values: List<String>): EndpointProfile {
```
Comment (sdhuka) [inline aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/targeting/endpointProfile/EndpointProfile.kt:154]: non null attribute name

File: aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/targeting/endpointProfile/EndpointProfile.kt [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+     */
+    val allMetrics: Map<String, Double>
+        get() = Collections.unmodifiableMap(metrics)
+
+    override fun toString(): String {
+        return toJSONObject().toString()
+    }
+
+    fun toJSONObject(): JsonObject {
+        return buildJsonObject {
```
Comment (sdhuka) [inline aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/targeting/endpointProfile/EndpointProfile.kt:262]: Can we replace this with kotlin json serializer?

File: aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/internal/core/util/StringUtil.kt [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+ * String utility methods.
+ */
+const val UTF_8 = "UTF-8"
+
+/**
+ * Determines if a string is null or zero-length.
+ *
+ * @return true if this string is null or zero-length, otherwise false
+ */
+fun String.isNullOrEmpty(): Boolean {
```
[the remark this answers is not in the document]
Comment (tylerjroach) [reply aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/internal/core/util/StringUtil.kt:37]: Why are we adding this? This already exists as extension function in Kotlin. Same for the other one in this class.

File: aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/targeting/TargetingClient.kt [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+internal class TargetingClient(
+    private val pinpointClient: PinpointClient,
+    pinpointNotificationClient: PinpointNotificationClient,
+    idService: SharedPrefsUniqueIdService,
+    private val prefs: SharedPreferences,
+    appDetails: AndroidAppDetails,
+    deviceDetails: AndroidDeviceDetails,
+    applicationContext: Context,
+) {
+    private val endpointProfile: EndpointProfile =
```
[the remark this answers is not in the document]
Comment (tjleing) [reply aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/targeting/TargetingClient.kt:55]: will be in a followup PR!

**Review by sdhuka — APPROVED** [review]

**Review by sdhuka — APPROVED** [review]

File: aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/targeting/endpointProfile/EndpointProfile.kt [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+                    try {
+                        put(key, JsonArray(value.map { JsonPrimitive(it) }))
+                    } catch (e: Exception) {
+                        // Do not log e due to potentially sensitive information
+                        LOG.warn("Error serializing attributes.")
+                    }
+                }
+            }
+            // If there are any attributes put then add the attributes to the structure
+            if (attributesJson.isEmpty()) {
```
Comment (tjleing) [inline aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/targeting/endpointProfile/EndpointProfile.kt:272]: I realized these were backwards while creating unit tests in a forthcoming PR -- I'll correct the logic there as well

**Review by gpanshu — APPROVED** [review]

**Pull request merged**